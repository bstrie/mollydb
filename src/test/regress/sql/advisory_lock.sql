--
-- ADVISORY LOCKS
--

BEGIN;

SELECT
	mdb_advisory_xact_lock(1), mdb_advisory_xact_lock_shared(2),
	mdb_advisory_xact_lock(1, 1), mdb_advisory_xact_lock_shared(2, 2);

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;


-- mdb_advisory_unlock_all() shouldn't release xact locks
SELECT mdb_advisory_unlock_all();

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';


-- can't unlock xact locks
SELECT
	mdb_advisory_unlock(1), mdb_advisory_unlock_shared(2),
	mdb_advisory_unlock(1, 1), mdb_advisory_unlock_shared(2, 2);


-- automatically release xact locks at commit
COMMIT;

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';


BEGIN;

-- holding both session and xact locks on the same objects, xact first
SELECT
	mdb_advisory_xact_lock(1), mdb_advisory_xact_lock_shared(2),
	mdb_advisory_xact_lock(1, 1), mdb_advisory_xact_lock_shared(2, 2);

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;

SELECT
	mdb_advisory_lock(1), mdb_advisory_lock_shared(2),
	mdb_advisory_lock(1, 1), mdb_advisory_lock_shared(2, 2);

ROLLBACK;

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;


-- unlocking session locks
SELECT
	mdb_advisory_unlock(1), mdb_advisory_unlock(1),
	mdb_advisory_unlock_shared(2), mdb_advisory_unlock_shared(2),
	mdb_advisory_unlock(1, 1), mdb_advisory_unlock(1, 1),
	mdb_advisory_unlock_shared(2, 2), mdb_advisory_unlock_shared(2, 2);

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';


BEGIN;

-- holding both session and xact locks on the same objects, session first
SELECT
	mdb_advisory_lock(1), mdb_advisory_lock_shared(2),
	mdb_advisory_lock(1, 1), mdb_advisory_lock_shared(2, 2);

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;

SELECT
	mdb_advisory_xact_lock(1), mdb_advisory_xact_lock_shared(2),
	mdb_advisory_xact_lock(1, 1), mdb_advisory_xact_lock_shared(2, 2);

ROLLBACK;

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;


-- releasing all session locks
SELECT mdb_advisory_unlock_all();

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';


BEGIN;

-- grabbing txn locks multiple times

SELECT
	mdb_advisory_xact_lock(1), mdb_advisory_xact_lock(1),
	mdb_advisory_xact_lock_shared(2), mdb_advisory_xact_lock_shared(2),
	mdb_advisory_xact_lock(1, 1), mdb_advisory_xact_lock(1, 1),
	mdb_advisory_xact_lock_shared(2, 2), mdb_advisory_xact_lock_shared(2, 2);

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;

COMMIT;

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';

-- grabbing session locks multiple times

SELECT
	mdb_advisory_lock(1), mdb_advisory_lock(1),
	mdb_advisory_lock_shared(2), mdb_advisory_lock_shared(2),
	mdb_advisory_lock(1, 1), mdb_advisory_lock(1, 1),
	mdb_advisory_lock_shared(2, 2), mdb_advisory_lock_shared(2, 2);

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;

SELECT
	mdb_advisory_unlock(1), mdb_advisory_unlock(1),
	mdb_advisory_unlock_shared(2), mdb_advisory_unlock_shared(2),
	mdb_advisory_unlock(1, 1), mdb_advisory_unlock(1, 1),
	mdb_advisory_unlock_shared(2, 2), mdb_advisory_unlock_shared(2, 2);

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';

-- .. and releasing them all at once

SELECT
	mdb_advisory_lock(1), mdb_advisory_lock(1),
	mdb_advisory_lock_shared(2), mdb_advisory_lock_shared(2),
	mdb_advisory_lock(1, 1), mdb_advisory_lock(1, 1),
	mdb_advisory_lock_shared(2, 2), mdb_advisory_lock_shared(2, 2);

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;

SELECT mdb_advisory_unlock_all();

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';
