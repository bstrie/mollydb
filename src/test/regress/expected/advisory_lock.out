--
-- ADVISORY LOCKS
--
BEGIN;
SELECT
	mdb_advisory_xact_lock(1), mdb_advisory_xact_lock_shared(2),
	mdb_advisory_xact_lock(1, 1), mdb_advisory_xact_lock_shared(2, 2);
 mdb_advisory_xact_lock | mdb_advisory_xact_lock_shared | mdb_advisory_xact_lock | mdb_advisory_xact_lock_shared 
-----------------------+------------------------------+-----------------------+------------------------------
                       |                              |                       | 
(1 row)

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;
 locktype | classid | objid | objsubid |     mode      | granted 
----------+---------+-------+----------+---------------+---------
 advisory |       0 |     1 |        1 | ExclusiveLock | t
 advisory |       0 |     2 |        1 | ShareLock     | t
 advisory |       1 |     1 |        2 | ExclusiveLock | t
 advisory |       2 |     2 |        2 | ShareLock     | t
(4 rows)

-- mdb_advisory_unlock_all() shouldn't release xact locks
SELECT mdb_advisory_unlock_all();
 mdb_advisory_unlock_all 
------------------------
 
(1 row)

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';
 count 
-------
     4
(1 row)

-- can't unlock xact locks
SELECT
	mdb_advisory_unlock(1), mdb_advisory_unlock_shared(2),
	mdb_advisory_unlock(1, 1), mdb_advisory_unlock_shared(2, 2);
WARNING:  you don't own a lock of type ExclusiveLock
WARNING:  you don't own a lock of type ShareLock
WARNING:  you don't own a lock of type ExclusiveLock
WARNING:  you don't own a lock of type ShareLock
 mdb_advisory_unlock | mdb_advisory_unlock_shared | mdb_advisory_unlock | mdb_advisory_unlock_shared 
--------------------+---------------------------+--------------------+---------------------------
 f                  | f                         | f                  | f
(1 row)

-- automatically release xact locks at commit
COMMIT;
SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';
 count 
-------
     0
(1 row)

BEGIN;
-- holding both session and xact locks on the same objects, xact first
SELECT
	mdb_advisory_xact_lock(1), mdb_advisory_xact_lock_shared(2),
	mdb_advisory_xact_lock(1, 1), mdb_advisory_xact_lock_shared(2, 2);
 mdb_advisory_xact_lock | mdb_advisory_xact_lock_shared | mdb_advisory_xact_lock | mdb_advisory_xact_lock_shared 
-----------------------+------------------------------+-----------------------+------------------------------
                       |                              |                       | 
(1 row)

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;
 locktype | classid | objid | objsubid |     mode      | granted 
----------+---------+-------+----------+---------------+---------
 advisory |       0 |     1 |        1 | ExclusiveLock | t
 advisory |       0 |     2 |        1 | ShareLock     | t
 advisory |       1 |     1 |        2 | ExclusiveLock | t
 advisory |       2 |     2 |        2 | ShareLock     | t
(4 rows)

SELECT
	mdb_advisory_lock(1), mdb_advisory_lock_shared(2),
	mdb_advisory_lock(1, 1), mdb_advisory_lock_shared(2, 2);
 mdb_advisory_lock | mdb_advisory_lock_shared | mdb_advisory_lock | mdb_advisory_lock_shared 
------------------+-------------------------+------------------+-------------------------
                  |                         |                  | 
(1 row)

ROLLBACK;
SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;
 locktype | classid | objid | objsubid |     mode      | granted 
----------+---------+-------+----------+---------------+---------
 advisory |       0 |     1 |        1 | ExclusiveLock | t
 advisory |       0 |     2 |        1 | ShareLock     | t
 advisory |       1 |     1 |        2 | ExclusiveLock | t
 advisory |       2 |     2 |        2 | ShareLock     | t
(4 rows)

-- unlocking session locks
SELECT
	mdb_advisory_unlock(1), mdb_advisory_unlock(1),
	mdb_advisory_unlock_shared(2), mdb_advisory_unlock_shared(2),
	mdb_advisory_unlock(1, 1), mdb_advisory_unlock(1, 1),
	mdb_advisory_unlock_shared(2, 2), mdb_advisory_unlock_shared(2, 2);
WARNING:  you don't own a lock of type ExclusiveLock
WARNING:  you don't own a lock of type ShareLock
WARNING:  you don't own a lock of type ExclusiveLock
WARNING:  you don't own a lock of type ShareLock
 mdb_advisory_unlock | mdb_advisory_unlock | mdb_advisory_unlock_shared | mdb_advisory_unlock_shared | mdb_advisory_unlock | mdb_advisory_unlock | mdb_advisory_unlock_shared | mdb_advisory_unlock_shared 
--------------------+--------------------+---------------------------+---------------------------+--------------------+--------------------+---------------------------+---------------------------
 t                  | f                  | t                         | f                         | t                  | f                  | t                         | f
(1 row)

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';
 count 
-------
     0
(1 row)

BEGIN;
-- holding both session and xact locks on the same objects, session first
SELECT
	mdb_advisory_lock(1), mdb_advisory_lock_shared(2),
	mdb_advisory_lock(1, 1), mdb_advisory_lock_shared(2, 2);
 mdb_advisory_lock | mdb_advisory_lock_shared | mdb_advisory_lock | mdb_advisory_lock_shared 
------------------+-------------------------+------------------+-------------------------
                  |                         |                  | 
(1 row)

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;
 locktype | classid | objid | objsubid |     mode      | granted 
----------+---------+-------+----------+---------------+---------
 advisory |       0 |     1 |        1 | ExclusiveLock | t
 advisory |       0 |     2 |        1 | ShareLock     | t
 advisory |       1 |     1 |        2 | ExclusiveLock | t
 advisory |       2 |     2 |        2 | ShareLock     | t
(4 rows)

SELECT
	mdb_advisory_xact_lock(1), mdb_advisory_xact_lock_shared(2),
	mdb_advisory_xact_lock(1, 1), mdb_advisory_xact_lock_shared(2, 2);
 mdb_advisory_xact_lock | mdb_advisory_xact_lock_shared | mdb_advisory_xact_lock | mdb_advisory_xact_lock_shared 
-----------------------+------------------------------+-----------------------+------------------------------
                       |                              |                       | 
(1 row)

ROLLBACK;
SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;
 locktype | classid | objid | objsubid |     mode      | granted 
----------+---------+-------+----------+---------------+---------
 advisory |       0 |     1 |        1 | ExclusiveLock | t
 advisory |       0 |     2 |        1 | ShareLock     | t
 advisory |       1 |     1 |        2 | ExclusiveLock | t
 advisory |       2 |     2 |        2 | ShareLock     | t
(4 rows)

-- releasing all session locks
SELECT mdb_advisory_unlock_all();
 mdb_advisory_unlock_all 
------------------------
 
(1 row)

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';
 count 
-------
     0
(1 row)

BEGIN;
-- grabbing txn locks multiple times
SELECT
	mdb_advisory_xact_lock(1), mdb_advisory_xact_lock(1),
	mdb_advisory_xact_lock_shared(2), mdb_advisory_xact_lock_shared(2),
	mdb_advisory_xact_lock(1, 1), mdb_advisory_xact_lock(1, 1),
	mdb_advisory_xact_lock_shared(2, 2), mdb_advisory_xact_lock_shared(2, 2);
 mdb_advisory_xact_lock | mdb_advisory_xact_lock | mdb_advisory_xact_lock_shared | mdb_advisory_xact_lock_shared | mdb_advisory_xact_lock | mdb_advisory_xact_lock | mdb_advisory_xact_lock_shared | mdb_advisory_xact_lock_shared 
-----------------------+-----------------------+------------------------------+------------------------------+-----------------------+-----------------------+------------------------------+------------------------------
                       |                       |                              |                              |                       |                       |                              | 
(1 row)

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;
 locktype | classid | objid | objsubid |     mode      | granted 
----------+---------+-------+----------+---------------+---------
 advisory |       0 |     1 |        1 | ExclusiveLock | t
 advisory |       0 |     2 |        1 | ShareLock     | t
 advisory |       1 |     1 |        2 | ExclusiveLock | t
 advisory |       2 |     2 |        2 | ShareLock     | t
(4 rows)

COMMIT;
SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';
 count 
-------
     0
(1 row)

-- grabbing session locks multiple times
SELECT
	mdb_advisory_lock(1), mdb_advisory_lock(1),
	mdb_advisory_lock_shared(2), mdb_advisory_lock_shared(2),
	mdb_advisory_lock(1, 1), mdb_advisory_lock(1, 1),
	mdb_advisory_lock_shared(2, 2), mdb_advisory_lock_shared(2, 2);
 mdb_advisory_lock | mdb_advisory_lock | mdb_advisory_lock_shared | mdb_advisory_lock_shared | mdb_advisory_lock | mdb_advisory_lock | mdb_advisory_lock_shared | mdb_advisory_lock_shared 
------------------+------------------+-------------------------+-------------------------+------------------+------------------+-------------------------+-------------------------
                  |                  |                         |                         |                  |                  |                         | 
(1 row)

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;
 locktype | classid | objid | objsubid |     mode      | granted 
----------+---------+-------+----------+---------------+---------
 advisory |       0 |     1 |        1 | ExclusiveLock | t
 advisory |       0 |     2 |        1 | ShareLock     | t
 advisory |       1 |     1 |        2 | ExclusiveLock | t
 advisory |       2 |     2 |        2 | ShareLock     | t
(4 rows)

SELECT
	mdb_advisory_unlock(1), mdb_advisory_unlock(1),
	mdb_advisory_unlock_shared(2), mdb_advisory_unlock_shared(2),
	mdb_advisory_unlock(1, 1), mdb_advisory_unlock(1, 1),
	mdb_advisory_unlock_shared(2, 2), mdb_advisory_unlock_shared(2, 2);
 mdb_advisory_unlock | mdb_advisory_unlock | mdb_advisory_unlock_shared | mdb_advisory_unlock_shared | mdb_advisory_unlock | mdb_advisory_unlock | mdb_advisory_unlock_shared | mdb_advisory_unlock_shared 
--------------------+--------------------+---------------------------+---------------------------+--------------------+--------------------+---------------------------+---------------------------
 t                  | t                  | t                         | t                         | t                  | t                  | t                         | t
(1 row)

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';
 count 
-------
     0
(1 row)

-- .. and releasing them all at once
SELECT
	mdb_advisory_lock(1), mdb_advisory_lock(1),
	mdb_advisory_lock_shared(2), mdb_advisory_lock_shared(2),
	mdb_advisory_lock(1, 1), mdb_advisory_lock(1, 1),
	mdb_advisory_lock_shared(2, 2), mdb_advisory_lock_shared(2, 2);
 mdb_advisory_lock | mdb_advisory_lock | mdb_advisory_lock_shared | mdb_advisory_lock_shared | mdb_advisory_lock | mdb_advisory_lock | mdb_advisory_lock_shared | mdb_advisory_lock_shared 
------------------+------------------+-------------------------+-------------------------+------------------+------------------+-------------------------+-------------------------
                  |                  |                         |                         |                  |                  |                         | 
(1 row)

SELECT locktype, classid, objid, objsubid, mode, granted
	FROM mdb_locks WHERE locktype = 'advisory'
	ORDER BY classid, objid, objsubid;
 locktype | classid | objid | objsubid |     mode      | granted 
----------+---------+-------+----------+---------------+---------
 advisory |       0 |     1 |        1 | ExclusiveLock | t
 advisory |       0 |     2 |        1 | ShareLock     | t
 advisory |       1 |     1 |        2 | ExclusiveLock | t
 advisory |       2 |     2 |        2 | ShareLock     | t
(4 rows)

SELECT mdb_advisory_unlock_all();
 mdb_advisory_unlock_all 
------------------------
 
(1 row)

SELECT count(*) FROM mdb_locks WHERE locktype = 'advisory';
 count 
-------
     0
(1 row)

